libvast/src/segment_builder.cpp

```diff
@@ -34,38 +35,42 @@ segment_builder::segment_builder() {
 caf::error segment_builder::add(table_slice_ptr x) {
   if (x->offset() < min_table_slice_offset_)
     return make_error(ec::unspecified, "slice offsets not increasing");
-  auto before = table_slice_buffer_.size();
-  caf::binary_serializer sink{nullptr, table_slice_buffer_};
-  if (auto error = sink(x)) {
-    table_slice_buffer_.resize(before);
-    return error;
-  }
-  auto after = table_slice_buffer_.size();
-  VAST_ASSERT(before < after);
-  meta_.slices.push_back({
-    detail::narrow_cast<int64_t>(before),
-    detail::narrow_cast<int64_t>(after),
-    x->offset(), x->rows()});
-  min_table_slice_offset_ = x->offset() + x->rows();
+  auto slice = fbs::create_table_slice(builder_, x);
+  if (!slice)
+    return slice.error();
+  table_slices_.push_back(*slice);
   slices_.push_back(x);
   return caf::none;
 }
 
 segment_ptr segment_builder::finish() {
-  if (meta_.slices.empty())
+  if (table_slices_.empty())
     return nullptr;
-  auto result = segment_ptr{new segment, false};
-  result->meta_ = std::move(meta_);
-  result->chunk_ = chunk::make(std::move(table_slice_buffer_));
-  result->header_.magic = segment::magic;
-  result->header_.version = segment::version;
-  result->header_.id = id_;
+  auto table_slices_offset = builder_.CreateVector(table_slices_);
+  auto uuid_offset = fbs::create_bytes(builder_, id_);
+  fbs::SegmentBuilder segment_builder{builder_};
+  segment_builder.add_data(table_slices_offset);
+  segment_builder.add_uuid(uuid_offset);
+  segment_builder.add_version(fbs::SegmentVersion::v1);
+  auto segment_offset = segment_builder.Finish();
+  builder_.Finish(segment_offset);
+  size_t offset;
+  size_t size;
+  auto ptr = builder_.ReleaseRaw(size, offset);
+#if 0
+  VAST_ASSERT(size > offset);
+  auto verifier = flatbuffers::Verifier{ptr + offset, size - offset};
+  VAST_ASSERT(fbs::VerifySegmentBuffer(verifier));
+#endif
+  auto deleter = [=]() { flatbuffers::DefaultAllocator::dealloc(ptr, size); };
+  auto chk = chunk::make(size - offset, ptr + offset, deleter);
```

Comment:
This means that the first `offset` bytes are used for flatbuffer internals?

Reply:
Yeah, there is some internal bookkeeping in the builder that needs to be skipped. Based on the documentation it's not clear exactly what that bookkeeping is, but Wouter gave green light when I presented this example on the mailing list.